File: 32.最长有效括号.py
#
# @lc app=leetcode.cn id=32 lang=python3
#
# [32] 最长有效括号
#

# @lc code=start
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        if not s or len(s) <= 1:
            return 0
    
        res = 0
        dp = [0 for _ in range(len(s))]
        for i in range(len(s)-2,-1,-1):
            tmp =  i + 1 + dp[i+1]
            if s[i] == '(' and tmp < len(s) and s[tmp] == ')':
                dp[i] = dp[i+1] + 2  # 把i和tmp并入 +2
                if tmp + 1 < len(s):
                    dp[i] += dp[tmp + 1]  # ")(())()"

                res = max(dp[i], res)
        return res




# @lc code=end

File: 32.最长有效括号.java
/*
 * @lc app=leetcode.cn id=32 lang=java
 *
 * [32] 最长有效括号
 */

// @lc code=start
class Solution {
    public int longestValidParentheses(String s) {
        if(s==null) return 0;
        int res = 0;
        int sLength = s.length();
        int[] dp = new int[sLength];
        for(int i=sLength-2; i>=0;i--){
            int tmp = dp[i+1] + i + 1;
            if(s.charAt(i)=='(' && tmp<sLength && s.charAt(tmp)==')' ){
                dp[i] = dp[i+1] + 2;
                if(tmp+1 < sLength) dp[i] += dp[tmp+1];


            }
            res = Math.max(res, dp[i]);
        }
        return res;
    }
}
// @lc code=end

